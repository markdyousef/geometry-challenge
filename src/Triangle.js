import React from "react";
import styled from "styled-components";

const TriangleContainer = styled.div``;

/**
 * Equilateral Triangle
 * - all 3 sides of equal length
 */
const Equilateral = styled.span`
  width: 0;
  height: 0;
  border-right: 80px solid transparent;
  border-bottom: 130px solid #444;
  border-left: 80px solid transparent;
`;

/**
 * Isosceles Triangle
 * - exactly 2 sides of equal lenght
 */
const Isosceles = styled.span`
  width: 0;
  height: 0;
  border-right: 80px solid transparent;
  border-bottom: 80px solid #000;
  border-left: 80px solid transparent;
`;

/**
 * Scalene Triangle
 * - no sides of equal length
 */
const Scalene = styled.span`
  width: 0;
  height: 0;
  border-right: 40px solid transparent;
  border-bottom: 100px solid #888;
  border-left: 150px solid transparent;
`;

export default class extends React.PureComponent {
  renderTriangle = type => {
    switch (type) {
      case "equilateral":
        return <Equilateral />;
      case "isosceles":
        return <Isosceles />;
      case "scalene":
        return <Scalene />;
      default:
        return null;
    }
  };
  render() {
    return (
      <TriangleContainer>
        {this.renderTriangle(this.props.type)}
      </TriangleContainer>
    );
  }
}
