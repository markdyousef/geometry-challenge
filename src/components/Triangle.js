import React from "react";
import styled from "styled-components";
import types from "../constants/triangle-types";
import Label from "./Label";
import { sidesToBorders } from "../lib/triangle-utils";

const MAX_WIDTH = 800;
const MAX_HEIGHT = 500;
const SCALE = 8;

const COLORS = [
  { type: types.EQUILATERAL, color: "#324D5C" },
  { type: types.ISOSCELES, color: "#9DC1DA" },
  { type: types.SCALENE, color: "#56A690" }
];

const Box = styled.div`
  background: #fff;
  border: 1px solid #ddd;
  width: ${MAX_WIDTH}px;
  height: ${MAX_HEIGHT}px;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
`;

const TriangleContainer = styled.div`
  position: relative;
`;
const Title = styled.h2`
  font-weight: 200;
  text-transform: lowercase;
`;
const NoTriangle = styled.div``;
const Triangle = styled.div`
  width: 0;
  height: 0;
  ${props => `
    border-left: ${props.left * SCALE}px solid transparent;
    border-bottom: ${props.height * SCALE}px solid ${props.color};
    border-right: ${props.right * SCALE}px solid transparent;
  `};
`;

// helper function to scale sides to fit inside box
const scale = sides => {
  const total = Object.values(sides).reduce((acc, curr) => acc + curr);
  const scaledSides = {
    a: (sides.a / total) * 100,
    b: (sides.b / total) * 100,
    c: (sides.c / total) * 100
  };
  const borders = sidesToBorders(scaledSides);
  return borders;
};
export default class extends React.PureComponent {
  createLabels = borders => {
    const aLabel = (
      <Label
        name="a"
        style={{
          left: `${(borders.left * SCALE) / 2 - 10}px`,
          top: `${(borders.height * SCALE) / 2}px`,
          position: "absolute"
        }}
        key="a_label"
      />
    );

    const cLabel = (
      <Label
        name="c"
        style={{
          right: `${(borders.right * SCALE) / 2 - 10}px`,
          top: `${(borders.height * SCALE) / 2}px`,
          position: "absolute"
        }}
        key="c_label"
      />
    );

    const bLabel = (
      <Label
        name="b"
        style={{
          bottom: "-15px",
          left: `${((borders.right + borders.left) * SCALE) / 2 - 10}px`,
          position: "absolute"
        }}
        key="b_label"
      />
    );
    return [aLabel, bLabel, cLabel];
  };
  renderTriangle = () => {
    const { type, sides, withLabels } = this.props;
    const borders = scale(sides);

    if (type) {
      const labels = withLabels ? this.createLabels(borders) : null;
      const color = COLORS.find(d => d.type === type).color;
      return (
        <TriangleContainer>
          {labels}
          <Triangle {...borders} color={color} />
        </TriangleContainer>
      );
    }
    return (
      <NoTriangle>
        <h5>Can't draw this shape...</h5>
      </NoTriangle>
    );
  };
  render() {
    return (
      <Box>
        <Title>{this.props.type}</Title>
        {this.renderTriangle()}
      </Box>
    );
  }
}
