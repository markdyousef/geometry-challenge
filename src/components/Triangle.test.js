import React from "react";
import { shallow } from "enzyme";
import Triangle from "./Triangle";

it("renders Triangle", () => {
  const wrapper = shallow(
    <Triangle type="EQUILATERAL" withLabels sides={{ a: 1, b: 1, c: 1 }} />
  );
  expect(wrapper).toMatchSnapshot();
});
