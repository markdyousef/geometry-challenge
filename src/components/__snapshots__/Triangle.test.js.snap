// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Triangle 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PureComponent
    sides={
        Object {
            "a": 1,
            "b": 1,
            "c": 1,
          }
    }
    type="EQUILATERAL"
    withLabels={true}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <styled.h2>
          EQUILATERAL
</styled.h2>,
        <styled.div>
          <PureComponent
                    name="a"
                    style={
                              Object {
                                        "left": "58px",
                                        "position": "absolute",
                                        "top": "116px",
                                      }
                    }
          />
          <PureComponent
                    name="b"
                    style={
                              Object {
                                        "bottom": "-15px",
                                        "left": "126px",
                                        "position": "absolute",
                                      }
                    }
          />
          <PureComponent
                    name="c"
                    style={
                              Object {
                                        "position": "absolute",
                                        "right": "58px",
                                        "top": "116px",
                                      }
                    }
          />
          <styled.div
                    color="#324D5C"
                    height={29}
                    left={17}
                    right={17}
          />
</styled.div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "EQUILATERAL",
        },
        "ref": null,
        "rendered": "EQUILATERAL",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            Array [
              <PureComponent
                name="a"
                style={
                                Object {
                                                "left": "58px",
                                                "position": "absolute",
                                                "top": "116px",
                                              }
                }
/>,
              <PureComponent
                name="b"
                style={
                                Object {
                                                "bottom": "-15px",
                                                "left": "126px",
                                                "position": "absolute",
                                              }
                }
/>,
              <PureComponent
                name="c"
                style={
                                Object {
                                                "position": "absolute",
                                                "right": "58px",
                                                "top": "116px",
                                              }
                }
/>,
            ],
            <styled.div
              color="#324D5C"
              height={29}
              left={17}
              right={17}
/>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "a_label",
            "nodeType": "class",
            "props": Object {
              "name": "a",
              "style": Object {
                "left": "58px",
                "position": "absolute",
                "top": "116px",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "b_label",
            "nodeType": "class",
            "props": Object {
              "name": "b",
              "style": Object {
                "bottom": "-15px",
                "left": "126px",
                "position": "absolute",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "c_label",
            "nodeType": "class",
            "props": Object {
              "name": "c",
              "style": Object {
                "position": "absolute",
                "right": "58px",
                "top": "116px",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "color": "#324D5C",
              "height": 29,
              "left": 17,
              "right": 17,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <styled.h2>
            EQUILATERAL
</styled.h2>,
          <styled.div>
            <PureComponent
                        name="a"
                        style={
                                    Object {
                                                "left": "58px",
                                                "position": "absolute",
                                                "top": "116px",
                                              }
                        }
            />
            <PureComponent
                        name="b"
                        style={
                                    Object {
                                                "bottom": "-15px",
                                                "left": "126px",
                                                "position": "absolute",
                                              }
                        }
            />
            <PureComponent
                        name="c"
                        style={
                                    Object {
                                                "position": "absolute",
                                                "right": "58px",
                                                "top": "116px",
                                              }
                        }
            />
            <styled.div
                        color="#324D5C"
                        height={29}
                        left={17}
                        right={17}
            />
</styled.div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "EQUILATERAL",
          },
          "ref": null,
          "rendered": "EQUILATERAL",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              Array [
                <PureComponent
                  name="a"
                  style={
                                    Object {
                                                      "left": "58px",
                                                      "position": "absolute",
                                                      "top": "116px",
                                                    }
                  }
/>,
                <PureComponent
                  name="b"
                  style={
                                    Object {
                                                      "bottom": "-15px",
                                                      "left": "126px",
                                                      "position": "absolute",
                                                    }
                  }
/>,
                <PureComponent
                  name="c"
                  style={
                                    Object {
                                                      "position": "absolute",
                                                      "right": "58px",
                                                      "top": "116px",
                                                    }
                  }
/>,
              ],
              <styled.div
                color="#324D5C"
                height={29}
                left={17}
                right={17}
/>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "a_label",
              "nodeType": "class",
              "props": Object {
                "name": "a",
                "style": Object {
                  "left": "58px",
                  "position": "absolute",
                  "top": "116px",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "b_label",
              "nodeType": "class",
              "props": Object {
                "name": "b",
                "style": Object {
                  "bottom": "-15px",
                  "left": "126px",
                  "position": "absolute",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "c_label",
              "nodeType": "class",
              "props": Object {
                "name": "c",
                "style": Object {
                  "position": "absolute",
                  "right": "58px",
                  "top": "116px",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "color": "#324D5C",
                "height": 29,
                "left": 17,
                "right": 17,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
